### PR TITLE
Misc updates

### PR DESCRIPTION
**Pull request by fredrikekre**
See commit messages. I can split into multiple PRs if that is preferred.

### CODE REVIEW

File: src/diagonal.jl [hunk 1]
```diff
@@ -6,35 +6,49 @@ function EmptyDiagonalPreconditioner(A::AbstractMatrix)
...
+        throw(ArgumentError("incompatible indices for input arguments"))
     end
-    return y
-end
-@inline function (\)(C::DiagonalPreconditioner, b)
-    y = zero(b)
-    @inbounds @simd for j ∈ 1:size(y, 2)
-        for i ∈ 1:length(C.D)
-            y[i,j] = b[i,j] / C.D[i]
+    @inbounds for (yj, bj) in zip(axes(y, 2), axes(b, 2))
```
Comment (mohdibntarek) [inline src/diagonal.jl:45]: Is this faster than using `ldiv!` for matrices? I know you improved the previous implementation but I am curious if we can do better.

Comment (mohdibntarek) [reply src/diagonal.jl:45]: I am talking about something like this
```julia
ldiv!(rand(2,2), Diagonal(rand(2)), rand(2,2))
```

Comment (mohdibntarek) [reply src/diagonal.jl:45]: I will merge this one and we can iterate on it in another PR if needed.

**Comment by mohdibntarek** [comment]
Thanks @fredrikekre !

File: src/diagonal.jl (hunk 1 above)
Comment (fredrikekre) [reply src/diagonal.jl:45]: I am not sure. This implementation only requires indexing to work so seems a bit easier, and `ldiv!` with a `Diagonal` probably does the same thing under the hood anyway.

**Pull request merged**